### PR TITLE
implementing s3 dlq writer

### PR DESCRIPTION
**Pull request by cmanning09**
### Description
Builds an S3 Dlq Writer for writing DLQObjects to an S3 bucket as part of #2298. Next PR will leverage the DLQ writer in the OpenSearch Sink. I will add a README file later as well to detail all the config options, metrics and functionality.
 
### Issues Resolved
part of #2298
 
### Check List
- [x] New functionality includes testing.
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-plugins/failures-common/src/main/java/org/opensearch/dataprepper/plugins/dlq/s3/S3DlqWriter.java [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+            .bucket(bucket)
+            .key(buildKey(pipelineName, pluginId))
+            .build();
+
+        final String content = deserialize(dlqObjects);
+
+        final PutObjectResponse response = timedPutObject(putObjectRequest, content);
+
+        if (!response.sdkHttpResponse().isSuccessful()) {
+            throw new IOException(String.format(
```
Comment (asifsmohammed) [inline data-prepper-plugins/failures-common/src/main/java/org/opensearch/dataprepper/plugins/dlq/s3/S3DlqWriter.java:97]: Adding metrics based on the response code here will be useful. You can do it in a different PR as an improvement.

File: data-prepper-plugins/failures-common/src/main/java/org/opensearch/dataprepper/plugins/dlq/s3/S3DlqWriter.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        this.bucket = s3DlqWriterConfig.getBucket();
+        this.keyPathPrefix = s3DlqWriterConfig.getKeyPathPrefix();
+        this.objectMapper = objectMapper;
+    }
+
+    @Override
+    public void write(final List<DlqObject> dlqObjects, final String pipelineName, final String pluginId) throws IOException {
+        try {
+            doWrite(dlqObjects, pipelineName, pluginId);
+            dlqS3RequestSuccessCounter.increment();
```
Comment (dinujoh) [inline data-prepper-plugins/failures-common/src/main/java/org/opensearch/dataprepper/plugins/dlq/s3/S3DlqWriter.java:77]: Do we need this ? We increment the counter based on dlq objects in next line.

**Review by asifsmohammed — CHANGES_REQUESTED** [review]

File: data-prepper-plugins/failures-common/src/main/java/org/opensearch/dataprepper/plugins/dlq/s3/S3DlqWriterConfig.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+            .build();
+    }
+
+    private void validateStsRoleArn() {
+        final Arn arn = getArn();
+        if (!"iam".equals(arn.service())) {
+            throw new IllegalArgumentException("sts_role_arn must be an IAM Role");
+        }
+        final Optional<String> resourceType = arn.resource().resourceType();
+        if (resourceType.isEmpty() || !resourceType.get().equals("role")) {
```
Comment (dinujoh) [inline data-prepper-plugins/failures-common/src/main/java/org/opensearch/dataprepper/plugins/dlq/s3/S3DlqWriterConfig.java:91]: move string to constant

File: data-prepper-plugins/failures-common/src/main/java/org/opensearch/dataprepper/plugins/dlq/s3/S3DlqWriterConfig.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+    private static final int MAX_NUMBER_OF_RETRIES = 3;
+    private static final String DEFAULT_AWS_REGION = "us-east-1";
+    @JsonProperty("bucket")
+    @NotNull
+    @Size(min = 3, max = 63, message = "bucket lengthy should be between 3 and 63 characters")
+    private String bucket;
+
+    @JsonProperty("key_path_prefix")
+    @Size(min = 1, max = 1024, message = " key_path_prefix length should be between 1 and 1024 characters")
```
Comment (dinujoh) [inline data-prepper-plugins/failures-common/src/main/java/org/opensearch/dataprepper/plugins/dlq/s3/S3DlqWriterConfig.java:40]: nit: extra space beginning of the message.

**Comment by dinujoh** [comment]
Changes look good. I see we are returning customer content in the exception. Should we have any option to mask that ?

**Review by dinujoh — DISMISSED** [review]

File: data-prepper-plugins/failures-common/src/main/java/org/opensearch/dataprepper/plugins/dlq/s3/S3DlqProvider.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    @DataPrepperPluginConstructor
+    public S3DlqProvider(final S3DlqWriterConfig s3DlqWriterConfig, final PluginMetrics pluginMetrics) {
+        Objects.requireNonNull(s3DlqWriterConfig);
+        this.s3DlqWriterConfig = s3DlqWriterConfig;
+        this.pluginMetrics = pluginMetrics;
+    }
+
+    @Override
+    public Optional<DlqWriter> getDlqWriter() {
+        final ObjectMapper objectMapper = new ObjectMapper();
```
[the remark this answers is not in the document]
Comment (cmanning09) [reply data-prepper-plugins/failures-common/src/main/java/org/opensearch/dataprepper/plugins/dlq/s3/S3DlqProvider.java:41]: I want to inject the objectMapper into the Writer so I can mock it for unit testing.

File: data-prepper-plugins/failures-common/src/main/java/org/opensearch/dataprepper/plugins/dlq/s3/S3DlqWriter.java [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+    private String deserialize(final List<DlqObject> dlqObjects) throws IOException {
+        try {
+            final String content = objectMapper.writeValueAsString(dlqObjects);
+
+            dlqS3RequestSizeBytesSummary.record(content.getBytes(StandardCharsets.UTF_8).length);
+
+            return content;
+        } catch (JsonProcessingException e) {
+            throw new IOException(String.format("Failed to build valid S3 request"));
```
[the remark this answers is not in the document]
Comment (cmanning09) [reply data-prepper-plugins/failures-common/src/main/java/org/opensearch/dataprepper/plugins/dlq/s3/S3DlqWriter.java:127]: The interface is generic and we are committing to throwing only IO exceptions. I should just wrap it. I will do that. Thanks for the feedback.

File: data-prepper-plugins/failures-common/src/main/java/org/opensearch/dataprepper/plugins/dlq/s3/S3DlqWriter.java
[the remark this answers is not in the document]
Comment (cmanning09) [reply data-prepper-plugins/failures-common/src/main/java/org/opensearch/dataprepper/plugins/dlq/s3/S3DlqWriter.java:1]: We are leveraging the batching that is already completed by the OpenSearch Sink. All failed requests from the opensearch sink batches will be aggregated and batched as part of DLQ.

**Comment by cmanning09** [comment]
> Changes look good. I see we are returning customer content in the exception. Should we have any option to mask that ?

I am masking these exceptions when I handle the logging in this PR: https://github.com/opensearch-project/data-prepper/pull/2429/files

File: data-prepper-plugins/failures-common/src/main/java/org/opensearch/dataprepper/plugins/dlq/s3/S3DlqWriter.java (hunk 2 above)
Comment (cmanning09) [reply data-prepper-plugins/failures-common/src/main/java/org/opensearch/dataprepper/plugins/dlq/s3/S3DlqWriter.java:77]: These are two different metrics. (ie Requests vs Records)

File: data-prepper-plugins/failures-common/src/main/java/org/opensearch/dataprepper/plugins/dlq/s3/S3DlqWriter.java [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+    @Override
+    public void write(final List<DlqObject> dlqObjects, final String pipelineName, final String pluginId) throws IOException {
+        try {
+            doWrite(dlqObjects, pipelineName, pluginId);
+            dlqS3RequestSuccessCounter.increment();
+            dlqS3RecordsSuccessCounter.increment(dlqObjects.size());
+        } catch (final Exception e) {
+            dlqS3RequestFailedCounter.increment();
+            dlqS3RecordsFailedCounter.increment(dlqObjects.size());
```
[the remark this answers is not in the document]
Comment (cmanning09) [reply data-prepper-plugins/failures-common/src/main/java/org/opensearch/dataprepper/plugins/dlq/s3/S3DlqWriter.java:81]: These are two different metrics. (ie Requests vs Records)

File: data-prepper-plugins/failures-common/src/main/java/org/opensearch/dataprepper/plugins/dlq/s3/S3DlqWriter.java [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+                "Failed to write to S3 dlq: [%s] to S3 due to status code: %d",
+                content, response.sdkHttpResponse().statusCode()));
+        }
+    }
+
+    private PutObjectResponse timedPutObject(final PutObjectRequest putObjectRequest, final String content) throws IOException {
+        try {
+            return dlqS3RequestTimer.recordCallable(() -> putObject(putObjectRequest, content));
+        } catch (final Exception ex) {
+            throw new IOException(String.format("Failed to write to S3 dlq: [%s] to S3.", content), ex);
```
[the remark this answers is not in the document]
Comment (cmanning09) [reply data-prepper-plugins/failures-common/src/main/java/org/opensearch/dataprepper/plugins/dlq/s3/S3DlqWriter.java:107]: Masking is happening this PR: https://github.com/opensearch-project/data-prepper/pull/2429

I can pull logging into here an mask that way. That would probably be better long term if this plugin is used else where.

File: data-prepper-plugins/failures-common/src/main/java/org/opensearch/dataprepper/plugins/dlq/DlqWriter.java [hunk 9]
```diff
@@ -25,7 +25,7 @@ public interface DlqWriter {
      *
      * @since 2.2
      */
-    void write(final List<DlqObject> dlqObjects) throws IOException;
+    void write(final List<DlqObject> dlqObjects, final String pipelineName, final String pluginId) throws IOException;
```
[the remark this answers is not in the document]
Comment (cmanning09) [reply data-prepper-plugins/failures-common/src/main/java/org/opensearch/dataprepper/plugins/dlq/DlqWriter.java:28]: This is pipeline name that the DLQ is associated with. It should be passed in as part of the loading the plugin. For the immediate use case with opensearch the pipeline name will be the pipeline name the opensearch sink receives.

File: data-prepper-plugins/failures-common/src/main/java/org/opensearch/dataprepper/plugins/dlq/s3/S3DlqWriter.java (hunk 1 above)
Comment (cmanning09) [reply data-prepper-plugins/failures-common/src/main/java/org/opensearch/dataprepper/plugins/dlq/s3/S3DlqWriter.java:97]: #2440 <- tracking here

File: data-prepper-plugins/failures-common/src/main/java/org/opensearch/dataprepper/plugins/dlq/s3/S3DlqWriter.java [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+public class S3DlqWriter implements DlqWriter {
+
+    static final String S3_DLQ_RECORDS_SUCCESS = "dlqS3RecordsSuccess";
+    static final String S3_DLQ_RECORDS_FAILED = "dlqS3RecordsFailed";
+    static final String S3_DLQ_REQUEST_SUCCESS = "dlqS3RequestSuccess";
+    static final String S3_DLQ_REQUEST_FAILED = "dlqS3RequestFailed";
+    static final String S3_DLQ_REQUEST_LATENCY = "dlqS3RequestLatency";
+    static final String S3_DLQ_REQUEST_SIZE_BYTES = "dlqS3RequestSizeBytes";
+    private static final String KEY_NAME_FORMAT = "dlq-v%s-%s-%s-%s-%s";
+    private static final String FULL_KEY_FORMAT = "%s/%s";
```
[the remark this answers is not in the document]
Comment (cmanning09) [reply data-prepper-plugins/failures-common/src/main/java/org/opensearch/dataprepper/plugins/dlq/s3/S3DlqWriter.java:44]: If they have a trailing path there will be an extra `/ `. The functionality still works it will just introduce an extra layer. I can add a validation to strip it but if someone already has a bucket with this prefix and wants to use it we would not be able to support it. I am inclined to leave it an provide documentation around it.

**Comment by dlvenable** [comment]
I think this is good.

I'm wondering if we should have this common failures project though. I don't think this is a solution to extend to other sinks.

For more robust solution would probably be to move the entire DLQ processing to a special Data Prepper pipeline for DLQ processing. This could be handled by data-prepper-core.

**Review by asifsmohammed — APPROVED** [review]
